File: 238.product-of-array-except-self.java
/*
 * @lc app=leetcode id=238 lang=java
 *
 * [238] Product of Array Except Self
 */

// @lc code=start
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ret = new int[nums.length];

        for (int ix = nums.length - 1; ix >= 0; --ix) {
            if (ix == nums.length - 1) ret[ix] = nums[ix];
            else ret[ix] = nums[ix] * ret[ix + 1];
        }
        
        for (int ix = 0; ix != nums.length; ++ix) {
            if (ix == 0) nums[ix] = nums[ix];
            else nums[ix] = nums[ix] * nums[ix - 1];
        }

        for (int ix = 0; ix != nums.length; ++ix) {
            if (ix == 0) ret[ix] = ret[ix + 1];
            else if (ix == nums.length - 1) ret[ix] = nums[nums.length - 2];
            else ret[ix] = nums[ix - 1] * ret[ix + 1];
        }
        return ret;
    }
}
// @lc code=end

File: 238.product-of-array-except-self.cpp
/*
 * @lc app=leetcode id=238 lang=cpp
 *
 * [238] Product of Array Except Self
 */

// @lc code=start

#include <algorithm>
#include <vector>
#include <numeric>
#include <functional>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int prod = 1;
        unsigned num_zeros = 0;
        decltype(nums.begin()) zero_pos;
        for (auto iter = nums.begin(); iter != nums.end(); ++iter) {
            if (*iter == 0) {
                zero_pos = iter; num_zeros += 1;
            } else {
                prod *= *iter;
            }
        }
        if (num_zeros > 1) return vector<int>(nums.size(), 0);
        if (num_zeros == 1) {
            fill(nums.begin(), nums.end(), 0);
            *zero_pos = prod;
            return nums;
        }
        transform(nums.begin(), nums.end(), nums.begin(), [&prod](const int &val) {
            return prod / val;
        });
        return nums;
    }
};
// @lc code=end
